### PR TITLE
Admin Category Management

### PR DESCRIPTION
**Pull request by aplazand20**
[Commands]
N/A

[Pre-conditions]
1) Must access as an Admin
2) Go to Manage Categories (its in the dropdown when you click the name of the current user)

[Expected Output]
1) See the Routes : Namespace
2) Admin can create a category 
3) Admin can show categories
4) Admin can edit category 
5) Admin can delete category

[Notes]
1) If the dropdown button won't work after one click or two, just refresh your browser

### CODE REVIEW

File: app/controllers/admin/categories_controller.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+          end
+      end
+
+      def edit
+        @category = Category.find(params[:id])
+      end
+
+      def update
+        @category = Category.find(params[:id])
+          if @category.update_attributes(category_params)
```
Comment (inahsanjuan) [inline app/controllers/admin/categories_controller.rb]: @shaniajubelag Indent this block backward

File: app/models/category.rb [hunk 2]
```diff
@@ -1,2 +1,5 @@
 class Category < ApplicationRecord
+    validates :title, presence: true, length: { maximum: 50 }
+    validates :description, presence: true, length: { maximum: 255 },
```
Comment (inahsanjuan) [inline app/models/category.rb]: @shaniajubelag Update to 2-spaced indention

File: app/views/admin/categories/_form.html.erb [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+        <%= f.text_field :title, class:'form-control' %>
+    </div>
+    <div class = 'form-group'>
+        <%= f.label :description %>
+        <%= f.text_area :description, class:'form-control'  %>
+    </div>
+    <p></p>
+<div class = 'form-group'>
+    <%= f.submit "Save", class: "btn btn-primary" %>
+</div>
```
Comment (inahsanjuan) [inline app/views/admin/categories/_form.html.erb]: @shaniajubelag Clean up your indentions properly

File: app/views/admin/categories/edit.html.erb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<% provide(:title, 'Edit category') %>
+<% provide(:button_text, 'Save changes' )%>
+
+<p></p>
+
+<h1>Update category</h1>
+    <div class= "row justify-content-center">
```
Comment (inahsanjuan) [inline app/views/admin/categories/edit.html.erb]: @shaniajubelag Indent backward

File: app/views/admin/categories/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,39 @@
+<% provide(:title, "Lessons") %>
+<p></p>
+<h1>All lessons</h1>
+
+    <div class="container">
+        <p></p>
+        <table class="table table-hover">
+            <thead>
+                <tr class="bg-primary">
+                    <th class="text-white" colspan="3">Category</th>
```
Comment (inahsanjuan) [inline app/views/admin/categories/index.html.erb]: @shaniajubelag Indent backward

File: app/views/admin/categories/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+<% provide(:title, 'Category')%>
+<p></p>
+<h1>Create category</h1>
+
+<div class="row justify-content-center">
+  <div class="col-md-6">
+        <%= render 'form' %>
```
Comment (inahsanjuan) [inline app/views/admin/categories/new.html.erb:7]: @shaniajubelag Should be 4-spaced, not 6

File: app/views/layouts/_header.html.erb [hunk 7]
```diff
@@ -20,21 +20,8 @@
...
-                                <li><%= link_to "Categories", lessons_path, class: "nav-link"  %></li>
-                                <li><%= link_to "Manage lessons", manage_lessons_path, class: "nav-link"  %></li>
-                            </ul>
-                        </li>
-                    <% else %>
-                        <li><%= link_to "Lessons", lessons_path, class: "nav-link"  %></li>
-                    <% end %>
-
+                <li><%= link_to "Lessons", lessons_path, class: "nav-link"  %>
+                </li>
```
Comment (inahsanjuan) [inline app/views/layouts/_header.html.erb]: @shaniajubelag This `</li>` should be in same line as line `23` element

File: app/views/static_pages/about.html.erb [hunk 8]
```diff
@@ -1,5 +1,5 @@
 <% provide(:title, "About") %>
-
+<p></p>
```
Comment (inahsanjuan) [inline app/views/static_pages/about.html.erb]: @shaniajubelag What is the purpose of the blank element here? Please remove
If you want to add space, just use margin utility classes in bootstrap, e.g. `mt-1`, `mt-2`, `mt-3`, `mt-4`, `mt-5`

File: app/views/static_pages/contact.html.erb [hunk 9]
```diff
@@ -1,4 +1,5 @@
 <% provide(:title, "Contact") %>
+<p></p>
```
Comment (inahsanjuan) [inline app/views/static_pages/contact.html.erb]: @shaniajubelag Same comment in https://github.com/framgia/ph_sels_rails_shania/pull/8/files#diff-39ce76d94a7eb2d6282c0ff8a4ce44a9R2

File: app/views/static_pages/help.html.erb [hunk 10]
```diff
@@ -1,4 +1,5 @@
 <% provide(:title, "Help") %>
+<p></p>
```
Comment (inahsanjuan) [inline app/views/static_pages/help.html.erb]: @shaniajubelag Same comment in https://github.com/framgia/ph_sels_rails_shania/pull/8/files#diff-39ce76d94a7eb2d6282c0ff8a4ce44a9R2

File: app/views/static_pages/home.html.erb [hunk 11]
```diff
@@ -1,4 +1,5 @@
 <div class="center jumbotron">
+<p></p>
```
Comment (inahsanjuan) [inline app/views/static_pages/home.html.erb]: @shaniajubelag Same comment in https://github.com/framgia/ph_sels_rails_shania/pull/8/files#diff-39ce76d94a7eb2d6282c0ff8a4ce44a9R2

File: app/views/static_pages/lessons.html.erb [hunk 12]
```diff
@@ -1,5 +1,32 @@
 <% provide(:title, "Lessons") %>
+<p></p>
```
Comment (inahsanjuan) [inline app/views/static_pages/lessons.html.erb]: @shaniajubelag Same comment in https://github.com/framgia/ph_sels_rails_shania/pull/8/files#diff-39ce76d94a7eb2d6282c0ff8a4ce44a9R2

File: app/views/static_pages/lessons.html.erb [hunk 13]
```diff
@@ -1,5 +1,32 @@
...
+<div class="row">
+  <div class="col-sm-6">
+    <div class="card">
+        <div class="card-header">Advance Lessons</div>
+        <div class="card-body text-secondary">
+        <p class="card-text">With supporting text below as a natural lead-in to additional content.</p>
+        <a href="#" class="btn btn-primary">Start</a>
+        </div>
+      </div>
+    </div>
```
Comment (inahsanjuan) [inline app/views/static_pages/lessons.html.erb]: @shaniajubelag Update to 4-spaced indention

File: app/views/users/edit.html.erb [hunk 14]
```diff
@@ -1,5 +1,6 @@
 <% provide(:title, 'Edit user') %>
 <% provide(:button_text, 'Save changes') %>
+<p></p>
```
Comment (inahsanjuan) [inline app/views/users/edit.html.erb]: @shaniajubelag Same comment in https://github.com/framgia/ph_sels_rails_shania/pull/8/files#diff-39ce76d94a7eb2d6282c0ff8a4ce44a9R2

File: app/views/users/index.html.erb [hunk 15]
```diff
@@ -1,12 +1,13 @@
 <% provide(:title, 'All users') %>
+<p></p>
```
Comment (inahsanjuan) [inline app/views/users/index.html.erb]: @shaniajubelag Same comment in https://github.com/framgia/ph_sels_rails_shania/pull/8/files#diff-39ce76d94a7eb2d6282c0ff8a4ce44a9R2

File: app/views/users/index.html.erb [hunk 16]
```diff
@@ -1,12 +1,13 @@
...
-    </ul>
+        <ul class="users">
+            <% @users.each do |user| %>
+                <%= render user %>
+            <% end %>
+        </ul>
 
-<%= will_paginate %>
\ No newline at end of file
+    <%= will_paginate %>
```
Comment (inahsanjuan) [inline app/views/users/index.html.erb]: @shaniajubelag Observe proper spacing. Indention is added when it is not needed, please indent backwards.

**Review by inahsanjuan — CHANGES_REQUESTED** [review]
- Non-admin users can still access this admin page `/admin/categories` : (screenshot) https://prnt.sc/oi7sbu

File: app/controllers/admin/categories_controller.rb (hunk 1 above)
Comment (aplazand20) [reply app/controllers/admin/categories_controller.rb]: @inahsanjuan I have a different folder for categories for the users

File: app/views/admin/categories/new.html.erb (hunk 6 above)
Comment (aplazand20) [reply app/views/admin/categories/new.html.erb:7]: Selection_022
It was already 4 in my VS Code.

Comment (aplazand20) [reply app/views/admin/categories/new.html.erb:7]: @inahsanjuan I pushed again but this file is still not outdated

File: app/views/admin/categories/_form.html.erb [hunk 17]
```diff
@@ -0,0 +1,15 @@
+<%= form_for [:admin, @category] do |f| %>
+<%= render 'shared/error_messages', object: @category %>
```
Comment (inahsanjuan) [inline app/views/admin/categories/_form.html.erb]: @shaniajubelag 
Indent this line forward

File: app/views/admin/categories/_form.html.erb [hunk 18]
```diff
@@ -0,0 +1,15 @@
...
+    <div class='form-group'>
+        <%= f.label :title %>
+        <%= f.text_field :title, class:'form-control' %>
+    </div>
+    <div class='form-group'>
+        <%= f.label :description %>
+        <%= f.text_area :description, class:'form-control'  %>
+    </div>
+    <p></p>
```
Comment (inahsanjuan) [inline app/views/admin/categories/_form.html.erb]: @shaniajubelag Remove empty element

File: app/views/admin/categories/edit.html.erb [hunk 19]
```diff
@@ -0,0 +1,11 @@
+<% provide(:title, 'Edit category') %>
+<% provide(:button_text, 'Save changes' )%>
+
+<p></p>
```
Comment (inahsanjuan) [inline app/views/admin/categories/edit.html.erb]: @shaniajubelag Remove empty element

File: app/views/admin/categories/index.html.erb [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+                        </td>
+                        <td>
+                            <%= link_to "Edit", edit_admin_category_url(category.id) %> | <%= link_to "Delete", admin_category_path(category.id), method: :delete,
+                            data: { confirm: "You sure?"} %>
+                        </td>
+                    </tr>
+                <% end %>
+            </tbody>
+        </table>
+    </div>
```
Comment (inahsanjuan) [inline app/views/admin/categories/index.html.erb]: @shaniajubelag Fix indentions in this file.
Please observe the `.container`, `thead` and `tbody` block elements

**Review by inahsanjuan — CHANGES_REQUESTED** [review]
- **Most `<p></p>` elements were removed, but the spacing was not added. Please try to add margin classes.**

**Pull request merged**